lib/api/core/storage/clientAdapter.js

```diff
@@ -0,0 +1,160 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2018 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const
+  errorsManager = require('../../../config/error-codes/throw')
+    .wrap('internal', 'storage_engine');
+
+/**
+ * Storage client adapter to perform validation on index/collection existence
+ * and to maintain the index/collection cache.
+ */
+class ClientAdapter {
+  /**
+   * @param {Elasticsearch} storageClient
+   * @param {IndexCache} indexCache
+   */
+  constructor (storageClient, indexCache) {
+    this._client = storageClient;
+    this._indexCache = indexCache;
+
+    // Methods that needs to assert index and collection existence
+    this._assertIndexAndCollectionMethods = [
+      'search',
+      'get',
+      'mGet',
+      'count',
+      'create',
+      'createOrReplace',
+      'update',
+      'replace',
+      'delete',
+      'deleteByQuery',
+      'getMapping',
+      'updateMapping',
+      'truncateCollection',
+      'import',
+      'refreshCollection',
+      'exists',
+      'mCreate',
+      'mCreateOrReplace',
+      'mUpdate',
+      'mReplace',
+      'mDelete'
+    ];
+
+    // Methods that needs to assert index existence
+    this._assertIndexMethods = [
+      'listCollections'
+    ];
+
+    // Methods directly bind to the storageClient
+    this._rawMethods = [
+      'scroll',
+      'listIndexes',
+      'listAliases',
+      'info',
+      'indexExists',
+      'collectionExists',
+      'init',
+      'createIndex'
+    ];
+
+    // Methods that need to assert index/collection existence and add/remove item
+    // from the indexCache
+    this._othersMethods = [
```

Comment:
Poor object name, since this appears to be the most important collection of methods.

What about `this._cacheChangeMethods`?